"use strict";(self.webpackChunkwebsite=self.webpackChunkwebsite||[]).push([[8757],{51100:(e,n,s)=>{s.r(n),s.d(n,{assets:()=>o,contentTitle:()=>r,default:()=>h,frontMatter:()=>i,metadata:()=>l,toc:()=>c});var a=s(85893),t=s(11151);const i={title:"ggplot2",description:"Learn how to plot with ggplot2 R package",tags:["Data Visualization","ggplot2","R","Charts","Visualization"]},r="Unlocking Data\u2019s Story: An Introduction to ggplot2",l={id:"visualization/ggplot2",title:"ggplot2",description:"Learn how to plot with ggplot2 R package",source:"@site/docs/visualization/ggplot2.md",sourceDirName:"visualization",slug:"/visualization/ggplot2",permalink:"/portfolio/docs/visualization/ggplot2",draft:!1,unlisted:!1,tags:[{label:"Data Visualization",permalink:"/portfolio/docs/tags/data-visualization"},{label:"ggplot2",permalink:"/portfolio/docs/tags/ggplot-2"},{label:"R",permalink:"/portfolio/docs/tags/r"},{label:"Charts",permalink:"/portfolio/docs/tags/charts"},{label:"Visualization",permalink:"/portfolio/docs/tags/visualization"}],version:"current",frontMatter:{title:"ggplot2",description:"Learn how to plot with ggplot2 R package",tags:["Data Visualization","ggplot2","R","Charts","Visualization"]},sidebar:"tutorialSidebar",previous:{title:"Data Visualization",permalink:"/portfolio/docs/visualization/"},next:{title:"lets-plot",permalink:"/portfolio/docs/visualization/lets-plot"}},o={},c=[{value:"What\u2019s Special About ggplot2?",id:"whats-special-about-ggplot2",level:2},{value:"Plot 2.8",id:"plot-28",level:2},{value:"Code Walkthrough",id:"code-walkthrough",level:3},{value:"Plot 3.13",id:"plot-313",level:2},{value:"Code Walkthrough",id:"code-walkthrough-1",level:3},{value:"Plot 4.7",id:"plot-47",level:2},{value:"Code Walkthrough",id:"code-walkthrough-2",level:3},{value:"Plot 4.9",id:"plot-49",level:2},{value:"Code Walkthrough",id:"code-walkthrough-3",level:3},{value:"Plot 4.10",id:"plot-410",level:2},{value:"Code Walkthrough",id:"code-walkthrough-4",level:3},{value:"Plot 5.4",id:"plot-54",level:2},{value:"Code Walkthrough",id:"code-walkthrough-5",level:3},{value:"Plot extra",id:"plot-extra",level:2},{value:"Code walkthrough",id:"code-walkthrough-6",level:3}];function d(e){const n={annotation:"annotation",code:"code",em:"em",h1:"h1",h2:"h2",h3:"h3",img:"img",li:"li",math:"math",mi:"mi",mn:"mn",mo:"mo",mrow:"mrow",msup:"msup",ol:"ol",p:"p",pre:"pre",semantics:"semantics",span:"span",strong:"strong",ul:"ul",...(0,t.a)(),...e.components};return(0,a.jsxs)(a.Fragment,{children:[(0,a.jsx)(n.h1,{id:"unlocking-datas-story-an-introduction-to-ggplot2",children:"Unlocking Data\u2019s Story: An Introduction to ggplot2"}),"\n",(0,a.jsxs)(n.p,{children:["Visualizing data can be the most engaging way to tell a story, reveal insights, and communicate complex information. When it comes to creating stunning, layered plots in R, ",(0,a.jsx)(n.strong,{children:"ggplot2"})," is the go-to tool. Part of the tidyverse collection, ggplot2 stands out not only for its flexibility and visual appeal but also for the grammar-based approach it offers. This grammar of graphics framework makes it intuitive to combine various components-like points, lines, shapes, and facets-to create richly informative plots."]}),"\n",(0,a.jsx)(n.h2,{id:"whats-special-about-ggplot2",children:"What\u2019s Special About ggplot2?"}),"\n",(0,a.jsx)(n.p,{children:"At its core, ggplot2 is structured around a series of layers. Each layer represents a different part of the plot, like data, geometries, and statistical transformations. This makes it incredibly flexible-you can start with a simple scatter plot and layer in colors, labels, and even transformations until you have exactly the visualization you need."}),"\n",(0,a.jsxs)(n.p,{children:["Another reason to love ggplot2 is its elegance with ",(0,a.jsx)(n.strong,{children:"data mapping"}),". Instead of manually coding each feature, you define a mapping of variables to visual properties, such as color, size, and shape. The rest is taken care of by ggplot2, which automates the hard work of scaling, positioning, and layering."]}),"\n",(0,a.jsx)(n.p,{children:"Let\u2019s look at ggplot2 in action through examples. Here, I'll walk through the visualizations I created for my PhD thesis, showing how you can build plots layer by layer to achieve impactful, publication-quality visuals."}),"\n",(0,a.jsx)(n.h2,{id:"plot-28",children:"Plot 2.8"}),"\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.img,{src:s(56761).Z+"",width:"3450",height:"3600"})}),"\n",(0,a.jsx)(n.pre,{children:(0,a.jsx)(n.code,{className:"language-r",metastring:'title="R"',children:'library(ggplot2)\nlibrary(reshape2)\nlibrary(ggpubr)\n\n# Set theme\ntheme_set(theme_bw())\n\n# Download datasets\nbase <- "https://raw.githubusercontent.com/smortezah/phd/refs/heads/main/plot/"\nfor (file in c("ir_hs_gg.tsv", "ir_diff_hs_gg.tsv", "ir_hs_gg.png")) {\n  download.file(paste0(base, "2.8/", file), file)\n}\n\n# Set input/output files\nfile_in_a <- "ir_hs_gg.tsv"\nfile_in_b <- "ir_diff_hs_gg.tsv"\nfile_out <- "ir_hs_gg.png"\n\n# Functions\nprepare_data <- function(filename, ir) {\n  df <- read.table(filename, header = TRUE)\n  df$ref <- rownames(df)\n\n  df_molten <-\n    melt(df, id.vars = "ref", variable.name = "tar", value.name = "nrc")\n  df_molten$ir <- ir\n\n  write.table(\n    df_molten,\n    file = paste(filename, ".tsv"),\n    quote = FALSE,\n    sep = "\\t",\n    col.names = NA\n  )\n}\n\n# Data\nhs_names_in <- c(\n  "HS1", "HS2", "HS3", "HS4", "HS5", "HS6", "HS7", "HS8", "HS9", "HS10",\n  "HS11", "HS12", "HS13", "HS14", "HS15", "HS16", "HS17", "HS18", "HS19",\n  "HS20", "HS21", "HS22", "HSX", "HSY", "HSAL", "HSMT", "HSUL", "HSUP"\n)\nhs_names_out <- c(\n  "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14",\n  "15", "16", "17", "18", "19", "20", "21", "22", "X", "Y", "AL", "MT", "UL",\n  "UP"\n)\ngg_names_in <- c(\n  "GG1", "GG2A", "GG2B", "GG3", "GG4", "GG5", "GG6", "GG7", "GG8", "GG9",\n  "GG10", "GG11", "GG12", "GG13", "GG14", "GG15", "GG16", "GG17", "GG18",\n  "GG19", "GG20", "GG21", "GG22", "GGX", "GGMT", "GGUL", "GGUP"\n)\ngg_names_out <- c(\n  "1", "2A", "2B", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13",\n  "14", "15", "16", "17", "18", "19", "20", "21", "22", "X", "MT", "UL", "UP"\n)\n\n## Data for a\ndf_hs_gg <- read.table(file_in_a, header = TRUE)\ndf_hs_gg$ref <-\n  factor(df_hs_gg$ref, levels = hs_names_in, labels = hs_names_out)\ndf_hs_gg$tar <-\n  factor(df_hs_gg$tar, levels = gg_names_in, labels = gg_names_out)\n\n## Data for b\ndf_diff_hs_gg <- read.table(file_in_b, header = TRUE)\ndf_diff_hs_gg$ref <-\n  factor(df_diff_hs_gg$ref, levels = hs_names_in, labels = hs_names_out)\ndf_diff_hs_gg$tar <-\n  factor(df_diff_hs_gg$tar, levels = gg_names_in, labels = gg_names_out)\n\n# Plot\na <- ggplot(df_hs_gg, aes(tar, ref)) +\n  geom_tile(aes(fill = nrc), colour = "white") +\n  coord_equal() +\n  scale_fill_gradient(name = "NRC", low = "steelblue", high = "white") +\n  ylab(expression(paste(italic("H. sapiens "), "(ref)"))) +\n  xlab(expression(paste(italic("G. gorilla "), "(tar)"))) +\n  facet_wrap(~ir.label, dir = "h")\n\nb <- ggplot(df_diff_hs_gg, aes(tar, ref)) +\n  geom_tile(aes(fill = nrc), colour = "white") +\n  coord_equal() +\n  scale_fill_gradient(\n    name = expression(NRC[0] ~ -~ NRC[1]),\n    low = "white",\n    high = "orange"\n  ) +\n  ylab(expression(paste(italic("H. sapiens "), "(ref)"))) +\n  xlab(expression(paste(italic("G. gorilla "), "(tar)")))\n\n# Save plot\nplot <- ggarrange(a, b, labels = c("a", "b"), nrow = 2)\nggsave(file_out, plot = plot, height = 12, width = 11.5)\n'})}),"\n",(0,a.jsx)(n.h3,{id:"code-walkthrough",children:"Code Walkthrough"}),"\n",(0,a.jsxs)(n.ol,{children:["\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsxs)(n.p,{children:[(0,a.jsx)(n.strong,{children:"Data Preparation"}),":"]}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:["We begin by downloading and reshaping the data into long format, suitable for ",(0,a.jsx)(n.code,{children:"ggplot2"})," visualization. Here, the datasets ",(0,a.jsx)(n.code,{children:"ir_hs_gg.tsv"})," and ",(0,a.jsx)(n.code,{children:"ir_diff_hs_gg.tsv"})," are tailored for comparison between human (",(0,a.jsx)(n.em,{children:"H. sapiens"}),") and gorilla (",(0,a.jsx)(n.em,{children:"G. gorilla"}),") NRC scores."]}),"\n",(0,a.jsx)(n.li,{children:'Custom mappings (from "HS1" to "1" and "GG1" to "1") standardize axis labels, making the data more readable and highlighting cross-species differences.'}),"\n"]}),"\n"]}),"\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsxs)(n.p,{children:[(0,a.jsx)(n.strong,{children:"Building the Panels"}),":"]}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:[(0,a.jsx)(n.strong,{children:"Panel (a)"}),': This panel displays NRC scores with and without IR conditions between human (ref) and gorilla (tar) with a color gradient from "steelblue" (low NRC) to "white" (high NRC), making it easy to interpret expression levels at a glance.']}),"\n",(0,a.jsxs)(n.li,{children:[(0,a.jsx)(n.strong,{children:"Panel (b)"}),': Here, we illustrate the difference in NRC values between IR conditions, with a gradient from "white" to "orange" emphasizing the intensity of NRC[0] - NRC[1]. This setup highlights regions where IR conditions have a stronger impact, providing an immediate sense of variation.']}),"\n"]}),"\n"]}),"\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsxs)(n.p,{children:[(0,a.jsx)(n.strong,{children:"Arranging the Panels"}),":"]}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:["Using ",(0,a.jsx)(n.code,{children:"ggarrange"}),', we display the two panels in a vertical layout with labeled subpanels ("a" and "b") to keep the comparison clear and structured.']}),"\n",(0,a.jsx)(n.li,{children:"Finally, the plot is saved as a PNG image with custom dimensions to ensure clarity for publication."}),"\n"]}),"\n"]}),"\n"]}),"\n",(0,a.jsx)(n.h2,{id:"plot-313",children:"Plot 3.13"}),"\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.img,{src:s(52947).Z+"",width:"2100",height:"900"})}),"\n",(0,a.jsx)(n.pre,{children:(0,a.jsx)(n.code,{className:"language-r",metastring:'title="R"',children:'library(ggplot2)\nlibrary(ggpubr)\nlibrary(desc)\n\n# Set theme\ntheme_set(theme_bw())\n\n# Download datasets\nbase <- "https://raw.githubusercontent.com/smortezah/phd/refs/heads/main/plot/"\ndownload.file(paste0(base, "3.13/ac_nc_size.tsv"), "ac_nc_size.tsv")\n\n# Set input/output files\nfile_in <- "ac_nc_size.tsv"\nfile_out <- "ac_nc.png"\n\n# Data\ndf <- read.table(file_in, header = TRUE)\nsorted_domain <- c("Viruses", "Archaea", "Bacteria", "Eukaryota")\n\ndf_ave <- data.frame(\n  domain = factor(sorted_domain, levels = sorted_domain),\n  ave.nc = c(0.96, 0.92, 0.92, 0.88),\n  color = c("grey50", "blue", "brown", "purple")\n)\n\n# Plot\na <- ggplot(df, aes(x = domain, y = nc)) +\n  geom_violin(aes(fill = domain)) +\n  scale_x_discrete(limits = sorted_domain) +\n  scale_fill_manual(\n    limits = sorted_domain,\n    values = c("grey50", "blue", "brown", "purple")\n  ) +\n  ylab("NC") +\n  theme(axis.title.x = element_blank(), legend.position = "none")\n\nb <-\n  ggplot(df_ave, aes(x = domain, y = ave.nc, label = ave.nc)) +\n  geom_segment(\n    aes(\n      x = domain,\n      xend = domain,\n      y = 0.85,\n      yend = ave.nc,\n    ),\n    linewidth = 0.75,\n    color = df_ave$color\n  ) +\n  geom_point(size = 15, color = df_ave$color) +\n  geom_text(color = "white", size = 3) +\n  ylim(0.85, 1) +\n  ylab("Average NC") +\n  theme(axis.title.x = element_blank())\n\n# Save plot\nplot <- ggarrange(a, b, labels = c("a", "b"))\nggsave(file_out, plot = plot, height = 3, width = 7)\n'})}),"\n",(0,a.jsx)(n.h3,{id:"code-walkthrough-1",children:"Code Walkthrough"}),"\n",(0,a.jsxs)(n.ol,{children:["\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsxs)(n.p,{children:[(0,a.jsx)(n.strong,{children:"Data Preparation"}),":"]}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:["After downloading the dataset ",(0,a.jsx)(n.code,{children:"ac_nc_size.tsv"}),", we load it into a dataframe (",(0,a.jsx)(n.code,{children:"df"}),") and define the order of the four biological domains (",(0,a.jsx)(n.code,{children:"sorted_domain"}),") for consistent sorting along the x-axis."]}),"\n",(0,a.jsxs)(n.li,{children:["We create ",(0,a.jsx)(n.code,{children:"df_ave"}),", a second dataframe for average NC values per domain. This dataframe also includes a color palette for each domain to distinguish them visually."]}),"\n"]}),"\n"]}),"\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsxs)(n.p,{children:[(0,a.jsx)(n.strong,{children:"Building the Panels"}),":"]}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:[(0,a.jsx)(n.strong,{children:"Panel (a)"}),":","\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsx)(n.li,{children:"This panel displays the distribution of NC values across domains using violin plots, which are ideal for visualizing the spread and density of data within categories."}),"\n",(0,a.jsxs)(n.li,{children:["We use ",(0,a.jsx)(n.code,{children:"scale_fill_manual"})," to set each domain's color, ensuring consistent hues across both panels."]}),"\n",(0,a.jsx)(n.li,{children:"The y-axis is labeled \u201cNC\u201d to indicate that this panel displays the range of NC values."}),"\n"]}),"\n"]}),"\n",(0,a.jsxs)(n.li,{children:[(0,a.jsx)(n.strong,{children:"Panel (b)"}),":","\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:["The second panel highlights the ",(0,a.jsx)(n.strong,{children:"average NC"})," per domain. Vertical segments from a base of 0.85 to the average NC level make it easy to locate each domain\u2019s position along the y-axis."]}),"\n",(0,a.jsx)(n.li,{children:"Each segment is followed by a large, colored point showing the average NC value, with a white label in the center displaying the exact value."}),"\n",(0,a.jsxs)(n.li,{children:["Custom colors from ",(0,a.jsx)(n.code,{children:"df_ave"})," ensure visual alignment with Panel (a), providing a cohesive look."]}),"\n"]}),"\n"]}),"\n"]}),"\n"]}),"\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsxs)(n.p,{children:[(0,a.jsx)(n.strong,{children:"Arranging and Saving the Plot"}),":"]}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:["Using ",(0,a.jsx)(n.code,{children:"ggarrange"}),", we combine Panels (a) and (b) horizontally, with labels for easy reference."]}),"\n",(0,a.jsx)(n.li,{children:"Finally, the plot is saved with a 7x3 aspect ratio, ideal for embedding in a publication or slide."}),"\n"]}),"\n"]}),"\n"]}),"\n",(0,a.jsx)(n.p,{children:"This layout offers an informative snapshot of NC distributions and averages by domain, making it easy to compare individual data variability with overall trends."}),"\n",(0,a.jsx)(n.h2,{id:"plot-47",children:"Plot 4.7"}),"\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.img,{src:s(81455).Z+"",width:"2100",height:"1350"})}),"\n",(0,a.jsx)(n.pre,{children:(0,a.jsx)(n.code,{className:"language-r",metastring:'title="R"',children:'library(ggplot2)\nlibrary(ggpubr)\nlibrary(dplyr)\n\n# Set theme\ntheme_set(theme_bw())\n\n# Set output file\nfile_out <- "cryfa_enc_aes.png"\n\n# Data\n## Data for time\ntime_data <- rbind(\n  c("Cryfa", 0.7, "FASTA", "Encrypt"),\n  c("Cryfa", 0.9, "FASTA", "Decrypt"),\n  c("Cryfa", 8.3, "FASTQ", "Encrypt"),\n  c("Cryfa", 8.2, "FASTQ", "Decrypt"),\n  c("AES Crypt", 3.2, "FASTA", "Encrypt"),\n  c("AES Crypt", 3.1, "FASTA", "Decrypt"),\n  c("AES Crypt", 19.2, "FASTQ", "Encrypt"),\n  c("AES Crypt", 16.8, "FASTQ", "Decrypt")\n)\n\ndf_time <- data.frame(\n  method = time_data[, 1],\n  time = as.numeric(time_data[, 2]),\n  format = time_data[, 3],\n  state = time_data[, 4]\n)\ndf_time <- df_time %>% mutate(time = ifelse(state == "Encrypt", -time, time))\n\n## Data for size\nsize_data <- rbind(\n  c("Cryfa", "FASTA", 2163, "Compressed", 6327, 2.9),\n  c("Cryfa", "FASTA", 4164, "Reduced", 6327, 2.9),\n  c("Cryfa", "FASTQ", 18706, "Compressed", 35542, 1.9),\n  c("Cryfa", "FASTQ", 16836, "Reduced", 35542, 1.9),\n  c("AES Crypt", "FASTA", 6327, "Compressed", 6327, 1.0),\n  c("AES Crypt", "FASTQ", 35542, "Compressed", 35542, 1.0)\n)\n\ndf_size <- data.frame(\n  method = size_data[, 1],\n  format = size_data[, 2],\n  size = as.numeric(size_data[, 3]) / 1024,\n  state = size_data[, 4],\n  total.size = as.numeric(size_data[, 5]) / 1024,\n  ratio = as.numeric(size_data[, 6])\n)\ndf_size$state <- reorder(df_size$state, df_size$size)\ndf_size <- df_size %>%\n  group_by(method, format) %>%\n  mutate(lab_ypos = cumsum(size) - 0.5 * size)\n\n# Plot\na <- ggplot(df_time, aes(x = method, y = time, fill = state)) +\n  geom_col() +\n  coord_flip() +\n  geom_text(\n    aes(\n      y = time,\n      label = ifelse(state == "Encrypt", -round(time, 1), round(time, 1))\n    ),\n    size = 3,\n    color = ifelse(\n      df_time$format == "FASTA" & df_time$method == "Cryfa",\n      "black",\n      "white"\n    ),\n    hjust = ifelse(\n      df_time$state == "Encrypt",\n      ifelse(df_time$format == "FASTA" & df_time$method == "Cryfa", 1.3, -0.3),\n      ifelse(df_time$format == "FASTA" & df_time$method == "Cryfa", -0.3, 1.3)\n    )\n  ) +\n  ylab("Time (min)") +\n  theme(axis.title.y = element_blank(), legend.position = "right") +\n  scale_fill_discrete(\n    name = "",\n    breaks = c("Encrypt", "Decrypt"),\n    labels = c(\n      "Encrypt /\\nCompact & Encrypt         ",\n      "Decrypt /\\nDecrypt & Unpack"\n    )\n  ) +\n  scale_y_continuous(breaks = c(0), labels = c(0)) +\n  facet_grid(format ~ ., scales = "free") +\n  geom_hline(yintercept = 0)\n\nb <- ggplot(df_size, aes(x = method, y = size, fill = state)) +\n  geom_blank(aes(x = method, y = size * 1.11)) +\n  geom_col() +\n  geom_text(\n    aes(y = lab_ypos, label = round(size, digits = 1)),\n    size = 3,\n    color = "white"\n  ) +\n  geom_text(\n    aes(y = total.size, label = paste0("CR = ", round(ratio, digits = 1))),\n    vjust = -0.75,\n    size = 2.5\n  ) +\n  theme(axis.title.x = element_blank(), legend.position = "right") +\n  ylab("Size (GB)") +\n  scale_fill_manual(\n    name = "",\n    breaks = c("Compressed", "Reduced"),\n    labels = c("Compressed", "Reduced\\n(Original - Compressed)"),\n    values = c("#E69F00", "#56B4E9")\n  ) +\n  facet_grid(. ~ format, scales = "free")\n\n# Save plot\nplot <- ggarrange(a, b, labels = c("a", "b"), nrow = 2)\nggsave(file_out, plot = plot, height = 4.5)\n'})}),"\n",(0,a.jsx)(n.h3,{id:"code-walkthrough-2",children:"Code Walkthrough"}),"\n",(0,a.jsxs)(n.ol,{children:["\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsxs)(n.p,{children:[(0,a.jsx)(n.strong,{children:"Data Preparation"}),":"]}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:[(0,a.jsx)(n.strong,{children:"Execution Time Data"}),": We load the encryption and decryption times for Cryfa and AES Crypt in both FASTA and FASTQ formats into ",(0,a.jsx)(n.code,{children:"df_time"}),". To make encryption and decryption times easily distinguishable on a centered bar chart, encryption times are given as negative values."]}),"\n",(0,a.jsxs)(n.li,{children:[(0,a.jsx)(n.strong,{children:"File Size Data"}),": In ",(0,a.jsx)(n.code,{children:"df_size"}),", we load compressed and reduced file sizes (in GB) for Cryfa and AES Crypt in both formats. Each row includes the total original file size, compressed file size, and calculated compression ratio (CR)."]}),"\n"]}),"\n"]}),"\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsxs)(n.p,{children:[(0,a.jsx)(n.strong,{children:"Building the Panels"}),":"]}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:[(0,a.jsx)(n.strong,{children:"Panel (a): Execution Time"}),":","\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:["We create a ",(0,a.jsx)(n.strong,{children:"horizontal bar chart"})," with each method\u2019s time split by state (encryption or decryption) for both formats, organized into separate facets for easy comparison."]}),"\n",(0,a.jsx)(n.li,{children:"Time labels are displayed within the bars. Custom color adjustments ensure readability, especially for Cryfa's FASTA data, where text is in black."}),"\n"]}),"\n"]}),"\n",(0,a.jsxs)(n.li,{children:[(0,a.jsx)(n.strong,{children:"Panel (b): File Size Comparison"}),":","\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:["This panel uses a ",(0,a.jsx)(n.strong,{children:"stacked bar chart"})," to represent compressed and reduced file sizes. Labels on each segment indicate the size in GB."]}),"\n",(0,a.jsxs)(n.li,{children:["Each bar\u2019s top displays the ",(0,a.jsx)(n.strong,{children:"Compression Ratio (CR)"}),", with ",(0,a.jsx)(n.code,{children:"CR = 1.0"})," for AES Crypt, indicating no additional compression, while Cryfa achieves reductions in size."]}),"\n",(0,a.jsx)(n.li,{children:"Custom colors highlight compressed (orange) and reduced (blue) states, and facets separate the data by format (FASTA and FASTQ)."}),"\n"]}),"\n"]}),"\n"]}),"\n"]}),"\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.strong,{children:"Arranging and Saving the Plot"})}),"\n"]}),"\n"]}),"\n",(0,a.jsx)(n.p,{children:"By combining time efficiency and compression effectiveness, Plot 4.7 visually conveys the advantages of Cryfa over AES Crypt for sequence data, making it clear where each tool excels."}),"\n",(0,a.jsx)(n.h2,{id:"plot-49",children:"Plot 4.9"}),"\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.img,{src:s(64801).Z+"",width:"2100",height:"2100"})}),"\n",(0,a.jsx)(n.pre,{children:(0,a.jsx)(n.code,{className:"language-r",metastring:'title="R"',children:'library(ggplot2)\nlibrary(ggpubr)\n\n# Set theme\ntheme_set(theme_bw())\n\n# Set output file\nfile_out <- "cryfa_thread.png"\n\n# Data\n# Data for time\ntime_data <- rbind(\n  c("viruses.fasta", 1, 0.16, "Real time", "enc"),\n  c("viruses.fasta", 1, 0.12, "Real time", "dec"),\n  c("viruses.fasta", 2, 0.08, "Real time", "enc"),\n  c("viruses.fasta", 2, 0.10, "Real time", "dec"),\n  c("viruses.fasta", 3, 0.06, "Real time", "enc"),\n  c("viruses.fasta", 3, 0.09, "Real time", "dec"),\n  c("viruses.fasta", 4, 0.06, "Real time", "enc"),\n  c("viruses.fasta", 4, 0.09, "Real time", "dec"),\n  c("viruses.fasta", 5, 0.05, "Real time", "enc"),\n  c("viruses.fasta", 5, 0.08, "Real time", "dec"),\n  c("viruses.fasta", 6, 0.05, "Real time", "enc"),\n  c("viruses.fasta", 6, 0.08, "Real time", "dec"),\n  c("viruses.fasta", 7, 0.04, "Real time", "enc"),\n  c("viruses.fasta", 7, 0.07, "Real time", "dec"),\n  c("viruses.fasta", 8, 0.04, "Real time", "enc"),\n  c("viruses.fasta", 8, 0.07, "Real time", "dec"),\n  #\n  c("viruses.fasta", 1, 0.14, "CPU time", "enc"),\n  c("viruses.fasta", 1, 0.12, "CPU time", "dec"),\n  c("viruses.fasta", 2, 0.14, "CPU time", "enc"),\n  c("viruses.fasta", 2, 0.12, "CPU time", "dec"),\n  c("viruses.fasta", 3, 0.15, "CPU time", "enc"),\n  c("viruses.fasta", 3, 0.13, "CPU time", "dec"),\n  c("viruses.fasta", 4, 0.17, "CPU time", "enc"),\n  c("viruses.fasta", 4, 0.14, "CPU time", "dec"),\n  c("viruses.fasta", 5, 0.17, "CPU time", "enc"),\n  c("viruses.fasta", 5, 0.15, "CPU time", "dec"),\n  c("viruses.fasta", 6, 0.19, "CPU time", "enc"),\n  c("viruses.fasta", 6, 0.15, "CPU time", "dec"),\n  c("viruses.fasta", 7, 0.19, "CPU time", "enc"),\n  c("viruses.fasta", 7, 0.16, "CPU time", "dec"),\n  c("viruses.fasta", 8, 0.20, "CPU time", "enc"),\n  c("viruses.fasta", 8, 0.16, "CPU time", "dec"),\n  #\n  c("DS-B1088_SR.fastq", 1, 0.72, "Real time", "enc"),\n  c("DS-B1088_SR.fastq", 1, 0.57, "Real time", "dec"),\n  c("DS-B1088_SR.fastq", 2, 0.42, "Real time", "enc"),\n  c("DS-B1088_SR.fastq", 2, 0.48, "Real time", "dec"),\n  c("DS-B1088_SR.fastq", 3, 0.33, "Real time", "enc"),\n  c("DS-B1088_SR.fastq", 3, 0.41, "Real time", "dec"),\n  c("DS-B1088_SR.fastq", 4, 0.31, "Real time", "enc"),\n  c("DS-B1088_SR.fastq", 4, 0.40, "Real time", "dec"),\n  c("DS-B1088_SR.fastq", 5, 0.28, "Real time", "enc"),\n  c("DS-B1088_SR.fastq", 5, 0.37, "Real time", "dec"),\n  c("DS-B1088_SR.fastq", 6, 0.26, "Real time", "enc"),\n  c("DS-B1088_SR.fastq", 6, 0.36, "Real time", "dec"),\n  c("DS-B1088_SR.fastq", 7, 0.25, "Real time", "enc"),\n  c("DS-B1088_SR.fastq", 7, 0.35, "Real time", "dec"),\n  c("DS-B1088_SR.fastq", 8, 0.24, "Real time", "enc"),\n  c("DS-B1088_SR.fastq", 8, 0.35, "Real time", "dec"),\n  #\n  c("DS-B1088_SR.fastq", 1, 0.59, "CPU time", "enc"),\n  c("DS-B1088_SR.fastq", 1, 0.50, "CPU time", "dec"),\n  c("DS-B1088_SR.fastq", 2, 0.63, "CPU time", "enc"),\n  c("DS-B1088_SR.fastq", 2, 0.55, "CPU time", "dec"),\n  c("DS-B1088_SR.fastq", 3, 0.65, "CPU time", "enc"),\n  c("DS-B1088_SR.fastq", 3, 0.56, "CPU time", "dec"),\n  c("DS-B1088_SR.fastq", 4, 0.72, "CPU time", "enc"),\n  c("DS-B1088_SR.fastq", 4, 0.62, "CPU time", "dec"),\n  c("DS-B1088_SR.fastq", 5, 0.76, "CPU time", "enc"),\n  c("DS-B1088_SR.fastq", 5, 0.64, "CPU time", "dec"),\n  c("DS-B1088_SR.fastq", 6, 0.78, "CPU time", "enc"),\n  c("DS-B1088_SR.fastq", 6, 0.68, "CPU time", "dec"),\n  c("DS-B1088_SR.fastq", 7, 0.81, "CPU time", "enc"),\n  c("DS-B1088_SR.fastq", 7, 0.70, "CPU time", "dec"),\n  c("DS-B1088_SR.fastq", 8, 0.83, "CPU time", "enc"),\n  c("DS-B1088_SR.fastq", 8, 0.72, "CPU time", "dec")\n)\n\ndf_time <- data.frame(\n  dataset = time_data[, 1],\n  thread = as.numeric(time_data[, 2]),\n  time = as.numeric(time_data[, 3]),\n  type = time_data[, 4],\n  state = time_data[, 5]\n)\ndf_time$thread <- factor(df_time$thread, levels = c(1:8))\n\n# Data for memory\nmemory_data <- rbind(\n  c("viruses.fasta", 1, 9, "Memory", "enc"),\n  c("viruses.fasta", 1, 9, "Memory", "dec"),\n  c("viruses.fasta", 2, 9, "Memory", "enc"),\n  c("viruses.fasta", 2, 13, "Memory", "dec"),\n  c("viruses.fasta", 3, 9, "Memory", "enc"),\n  c("viruses.fasta", 3, 15, "Memory", "dec"),\n  c("viruses.fasta", 4, 9, "Memory", "enc"),\n  c("viruses.fasta", 4, 15, "Memory", "dec"),\n  c("viruses.fasta", 5, 9, "Memory", "enc"),\n  c("viruses.fasta", 5, 18, "Memory", "dec"),\n  c("viruses.fasta", 6, 9, "Memory", "enc"),\n  c("viruses.fasta", 6, 19, "Memory", "dec"),\n  c("viruses.fasta", 7, 9, "Memory", "enc"),\n  c("viruses.fasta", 7, 23, "Memory", "dec"),\n  c("viruses.fasta", 8, 11, "Memory", "enc"),\n  c("viruses.fasta", 8, 25, "Memory", "dec"),\n  #\n  c("DS-B1088_SR.fastq", 1, 11, "Memory", "enc"),\n  c("DS-B1088_SR.fastq", 1, 10, "Memory", "dec"),\n  c("DS-B1088_SR.fastq", 2, 11, "Memory", "enc"),\n  c("DS-B1088_SR.fastq", 2, 13, "Memory", "dec"),\n  c("DS-B1088_SR.fastq", 3, 11, "Memory", "enc"),\n  c("DS-B1088_SR.fastq", 3, 16, "Memory", "dec"),\n  c("DS-B1088_SR.fastq", 4, 11, "Memory", "enc"),\n  c("DS-B1088_SR.fastq", 4, 18, "Memory", "dec"),\n  c("DS-B1088_SR.fastq", 5, 11, "Memory", "enc"),\n  c("DS-B1088_SR.fastq", 5, 23, "Memory", "dec"),\n  c("DS-B1088_SR.fastq", 6, 11, "Memory", "enc"),\n  c("DS-B1088_SR.fastq", 6, 26, "Memory", "dec"),\n  c("DS-B1088_SR.fastq", 7, 11, "Memory", "enc"),\n  c("DS-B1088_SR.fastq", 7, 28, "Memory", "dec"),\n  c("DS-B1088_SR.fastq", 8, 12, "Memory", "enc"),\n  c("DS-B1088_SR.fastq", 8, 31, "Memory", "dec")\n)\n\ndf_memory <- data.frame(\n  dataset = memory_data[, 1],\n  thread = as.numeric(memory_data[, 2]),\n  memory = as.numeric(memory_data[, 3]),\n  type = memory_data[, 4],\n  state = memory_data[, 5]\n)\ndf_memory$thread <- factor(df_memory$thread, levels = c(1:8))\n\n# Plot\na <- ggplot(df_time, aes(thread, time * 60)) +\n  geom_linerange(\n    aes(x = thread, ymin = 0, ymax = time * 60, group = state),\n    color = "lightgray", linewidth = 1.0,\n    position = position_dodge2(0.6, reverse = TRUE)\n  ) +\n  geom_point(\n    aes(color = state),\n    position = position_dodge2(0.6, reverse = TRUE),\n    size = 3.75\n  ) +\n  facet_grid(type ~ dataset) +\n  xlab("Number of threads") +\n  ylab("Time (sec)") +\n  scale_color_discrete(\n    breaks = c("enc", "dec"),\n    labels = c("Compact &\\nEncrypt", "Decrypt &\\nUnpack")\n  ) +\n  theme(legend.title = element_blank())\n\nb <- ggplot(df_memory, aes(thread, memory)) +\n  geom_linerange(\n    aes(x = thread, ymin = 0, ymax = memory, group = state),\n    color = "lightgray", linewidth = 1.0,\n    position = position_dodge2(0.6, reverse = TRUE)\n  ) +\n  geom_point(\n    aes(color = state),\n    position = position_dodge2(0.6, reverse = TRUE),\n    size = 3.75\n  ) +\n  facet_grid(~dataset) +\n  xlab("Number of threads") +\n  ylab("Memory (MB)") +\n  scale_color_discrete(\n    breaks = c("enc", "dec"),\n    labels = c("Compact &\\nEncrypt", "Decrypt &\\nUnpack")\n  ) +\n  theme(legend.title = element_blank())\n\n# Save plot\nplot <- ggarrange(a, b, labels = c("a", "b"), nrow = 2, heights = c(1.85, 1))\nggsave(file_out, plot = plot)\n'})}),"\n",(0,a.jsx)(n.h3,{id:"code-walkthrough-3",children:"Code Walkthrough"}),"\n",(0,a.jsxs)(n.ol,{children:["\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsxs)(n.p,{children:[(0,a.jsx)(n.strong,{children:"Data Preparation"}),":"]}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:["Two data frames (",(0,a.jsx)(n.code,{children:"df_time"})," and ",(0,a.jsx)(n.code,{children:"df_memory"}),") are set up with performance data across threads."]}),"\n",(0,a.jsxs)(n.li,{children:[(0,a.jsx)(n.code,{children:"df_time"})," contains timing data (both real and CPU times in seconds) for each dataset, type of operation (encryption or decryption), and thread count."]}),"\n",(0,a.jsxs)(n.li,{children:[(0,a.jsx)(n.code,{children:"df_memory"})," provides memory usage for the same operations across thread counts."]}),"\n"]}),"\n"]}),"\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsxs)(n.p,{children:[(0,a.jsx)(n.strong,{children:"Plotting"}),":"]}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:["Two main plots are created and then combined vertically:","\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:[(0,a.jsx)(n.strong,{children:"Time Plot"})," (",(0,a.jsx)(n.code,{children:"a"}),"): Shows encryption and decryption times by thread count. Each point\u2019s position reflects the time in seconds, and lineranges represent these values. The plot is faceted by dataset and time type (real/CPU), with labels for each operation."]}),"\n",(0,a.jsxs)(n.li,{children:[(0,a.jsx)(n.strong,{children:"Memory Plot"})," (",(0,a.jsx)(n.code,{children:"b"}),"): Displays memory usage for each dataset and thread count. Similar to the time plot, lineranges and points represent memory values, and labels differentiate operations."]}),"\n"]}),"\n"]}),"\n",(0,a.jsx)(n.li,{children:"Both plots use custom color schemes and legend formatting to clarify encryption and decryption tasks."}),"\n"]}),"\n"]}),"\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.strong,{children:"Saving the Plot"})}),"\n"]}),"\n"]}),"\n",(0,a.jsx)(n.h2,{id:"plot-410",children:"Plot 4.10"}),"\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.img,{src:s(87197).Z+"",width:"2100",height:"2850"})}),"\n",(0,a.jsx)(n.pre,{children:(0,a.jsx)(n.code,{className:"language-r",metastring:'title="R"',children:'library(ggplot2)\nlibrary(dplyr)\nlibrary(ggpubr)\n\n# Set theme\ntheme_set(theme_bw())\n\n# Download datasets\nbase <- "https://raw.githubusercontent.com/smortezah/phd/refs/heads/main/plot/"\ndownload.file(paste0(base, "4.10/cryfa_redun.tsv"), "cryfa_redun.tsv")\n\n# Set input/output files\nfile_in <- "cryfa_redun.tsv"\nfile_out <- "cryfa_redun.png"\n\n# Data and plot\ndf <- read.table(file_in, header = TRUE)\n\n## 1.1\ncryfa <- df %>% filter(Method == "cryfa")\ndeliminate <- df %>% filter(Method == "DELIMINATE")\nmfcompress <- df %>% filter(Method == "MFCompress")\n\ncolor_list <- c("blue", "purple", "red", "darkgreen", "darkgray")\n\ng11 <- ggplot(cryfa, aes(x = Size / 1024 / 1024, y = NC, color = Category)) +\n  geom_point() +\n  scale_color_manual(\n    name = "",\n    breaks = c("A", "B", "F", "P", "V"),\n    labels = c("Archaea", "Bacteria", "Fungi", "Plants", "Viruses"),\n    values = color_list\n  ) +\n  scale_y_continuous(\n    limits = c(1.3, 1.5),\n    breaks = c(1.3, 1.4, 1.5),\n    labels = c(1.3, 1.4, 1.5)\n  ) +\n  xlim(0, 2) +\n  theme(axis.title.x = element_blank()) +\n  ggtitle("Cryfa") +\n  theme(legend.position = "top")\n\n## 1.2\narchaea <- subset(cryfa, Category == "A")\nbacteria <- subset(cryfa, Category == "B")\nfungi <- subset(cryfa, Category == "F")\nplants <- subset(cryfa, Category == "P")\nviruses <- subset(cryfa, Category == "V")\ndf_pointrange <- data.frame(\n  Category = c("A", "B", "F", "P", "V"),\n  lower = c(\n    min(archaea[, "NC"]),\n    min(bacteria[, "NC"]),\n    min(fungi[, "NC"]),\n    min(plants[, "NC"]),\n    min(viruses[, "NC"])\n  ),\n  mean = c(\n    mean(archaea[, "NC"]),\n    mean(bacteria[, "NC"]),\n    mean(fungi[, "NC"]),\n    mean(plants[, "NC"]),\n    mean(viruses[, "NC"])\n  ),\n  upper = c(\n    max(archaea[, "NC"]),\n    max(bacteria[, "NC"]),\n    max(fungi[, "NC"]),\n    max(plants[, "NC"]),\n    max(viruses[, "NC"])\n  )\n)\n\ng12 <- ggplot(df_pointrange) +\n  geom_pointrange(\n    aes(x = Category, ymin = lower, y = mean, ymax = upper, color = Category),\n    fill = "white",\n    fatten = 4,\n    shape = 23\n  ) +\n  theme(axis.title.x = element_blank(), axis.title.y = element_blank()) +\n  scale_color_manual(values = color_list) +\n  scale_fill_manual(values = color_list) +\n  scale_y_continuous(\n    position = "left",\n    limits = c(1.3, 1.5),\n    breaks = c(1.3, 1.4, 1.5),\n    labels = c(1.3, 1.4, 1.5)\n  ) +\n  ggtitle("Cryfa")\n\n## 2.1\ng21 <-\n  ggplot(deliminate, aes(x = Size / 1024 / 1024, y = NC, color = Category)) +\n  geom_point() +\n  scale_color_manual(values = color_list) +\n  xlim(0, 2) +\n  ylim(0.7, 1.1) +\n  theme(axis.title.x = element_blank()) +\n  ggtitle("DELIMINATE")\n\n## 2.2\narchaea <- subset(deliminate, Category == "A")\nbacteria <- subset(deliminate, Category == "B")\nfungi <- subset(deliminate, Category == "F")\nplants <- subset(deliminate, Category == "P")\nviruses <- subset(deliminate, Category == "V")\ndf_pointrange <- data.frame(\n  Category = c("A", "B", "F", "P", "V"),\n  lower = c(\n    min(archaea[, "NC"]),\n    min(bacteria[, "NC"]),\n    min(fungi[, "NC"]),\n    min(plants[, "NC"]),\n    min(viruses[, "NC"])\n  ),\n  mean = c(\n    mean(archaea[, "NC"]),\n    mean(bacteria[, "NC"]),\n    mean(fungi[, "NC"]),\n    mean(plants[, "NC"]),\n    mean(viruses[, "NC"])\n  ),\n  upper = c(\n    max(archaea[, "NC"]),\n    max(bacteria[, "NC"]),\n    max(fungi[, "NC"]),\n    max(plants[, "NC"]),\n    max(viruses[, "NC"])\n  )\n)\n\ng22 <- ggplot(df_pointrange) +\n  geom_pointrange(\n    aes(x = Category, ymin = lower, y = mean, ymax = upper, color = Category),\n    fill = "white",\n    fatten = 4,\n    shape = 23\n  ) +\n  theme(axis.title.x = element_blank(), axis.title.y = element_blank()) +\n  scale_color_manual(values = color_list) +\n  scale_fill_manual(values = color_list) +\n  xlab("") +\n  scale_y_continuous(position = "left", limits = c(0.7, 1.1)) +\n  ggtitle("DELIMINATE")\n\n## 3.1\ng31 <-\n  ggplot(mfcompress, aes(x = Size / 1024 / 1024, y = NC, color = Category)) +\n  geom_point() +\n  scale_color_manual(values = color_list) +\n  xlim(0, 2) +\n  ylim(0.7, 1.1) +\n  xlab("File size (MB)") +\n  ggtitle("MFCompress")\n\n## 3.2\narchaea <- subset(mfcompress, Category == "A")\nbacteria <- subset(mfcompress, Category == "B")\nfungi <- subset(mfcompress, Category == "F")\nplants <- subset(mfcompress, Category == "P")\nviruses <- subset(mfcompress, Category == "V")\ndf_pointrange <- data.frame(\n  Category = c("A", "B", "F", "P", "V"),\n  lower = c(\n    min(archaea[, "NC"]),\n    min(bacteria[, "NC"]),\n    min(fungi[, "NC"]),\n    min(plants[, "NC"]),\n    min(viruses[, "NC"])\n  ),\n  mean = c(\n    mean(archaea[, "NC"]),\n    mean(bacteria[, "NC"]),\n    mean(fungi[, "NC"]),\n    mean(plants[, "NC"]),\n    mean(viruses[, "NC"])\n  ),\n  upper = c(\n    max(archaea[, "NC"]),\n    max(bacteria[, "NC"]),\n    max(fungi[, "NC"]),\n    max(plants[, "NC"]),\n    max(viruses[, "NC"])\n  )\n)\n\ng32 <- ggplot(df_pointrange) +\n  geom_pointrange(\n    aes(x = Category, ymin = lower, y = mean, ymax = upper, color = Category),\n    fill = "white",\n    fatten = 4,\n    shape = 23\n  ) +\n  theme(axis.title.y = element_blank()) +\n  scale_color_manual(values = color_list) +\n  scale_fill_manual(values = color_list) +\n  xlab("") +\n  scale_y_continuous(position = "left", limits = c(0.7, 1.1)) +\n  ggtitle("MFCompress")\n\n# Save plot\nplot <- ggarrange(\n  g11, g12,\n  g21, g22,\n  g31, g32,\n  labels = c("a", "b"),\n  nrow = 3,\n  ncol = 2,\n  widths = c(4, 1),\n  heights = c(1.2, 2, 2),\n  common.legend = TRUE\n)\nggsave(file_out, plot = plot, height = 9.5)\n'})}),"\n",(0,a.jsx)(n.h3,{id:"code-walkthrough-4",children:"Code Walkthrough"}),"\n",(0,a.jsxs)(n.ol,{children:["\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.strong,{children:"Data Preparation and Filtering:"})}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:["The dataset is loaded and filtered to separate rows by each compression method (",(0,a.jsx)(n.code,{children:"cryfa"}),", ",(0,a.jsx)(n.code,{children:"deliminate"}),", ",(0,a.jsx)(n.code,{children:"mfcompress"}),")."]}),"\n",(0,a.jsx)(n.li,{children:"Each method is associated with biological categories (e.g., Archaea, Bacteria) and analyzed separately."}),"\n"]}),"\n"]}),"\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.strong,{children:"Scatter Plots of Compression Methods:"})}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:["For each method (",(0,a.jsx)(n.code,{children:"g11"})," for Cryfa, ",(0,a.jsx)(n.code,{children:"g21"})," for DELIMINATE, ",(0,a.jsx)(n.code,{children:"g31"})," for MFCompress), a scatter plot is created."]}),"\n",(0,a.jsx)(n.li,{children:"The x-axis represents the file size in megabytes, and the y-axis shows the NC values (which likely represent compression efficiency)."}),"\n",(0,a.jsx)(n.li,{children:"Custom color coding is used to distinguish between categories, and legend labels are customized."}),"\n"]}),"\n"]}),"\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.strong,{children:"Point Range Summaries by Category:"})}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:["For each method, a ",(0,a.jsx)(n.code,{children:"geom_pointrange"})," plot is added to visualize the minimum, mean, and maximum NC values for each category."]}),"\n",(0,a.jsxs)(n.li,{children:["These plots (e.g., ",(0,a.jsx)(n.code,{children:"g12"}),", ",(0,a.jsx)(n.code,{children:"g22"}),", ",(0,a.jsx)(n.code,{children:"g32"}),") highlight the spread and average NC values for each category within the compression method."]}),"\n"]}),"\n"]}),"\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.strong,{children:"Plot Arrangement and Saving:"})}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsx)(n.li,{children:"The six plots are arranged in a 3x2 grid with labeled panels (a, b, etc.), aligning side-by-side the scatter and point range for each method."}),"\n",(0,a.jsxs)(n.li,{children:["The final plot is saved as a PNG file (",(0,a.jsx)(n.code,{children:"cryfa_redun.png"}),")."]}),"\n"]}),"\n"]}),"\n"]}),"\n",(0,a.jsx)(n.h2,{id:"plot-54",children:"Plot 5.4"}),"\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.img,{src:s(88847).Z+"",width:"2100",height:"2100"})}),"\n",(0,a.jsx)(n.pre,{children:(0,a.jsx)(n.code,{className:"language-r",metastring:'title="R"',children:'library(ggplot2)\nlibrary(ggridges)\nlibrary(ggpubr)\n\n# Set theme\ntheme_set(theme_bw())\n\n# Download datasets\nbase <- "https://raw.githubusercontent.com/smortezah/phd/refs/heads/main/plot/"\nfor (file in c("fruit_hs_uniq.tsv", "fruit_hs_stat.tsv")) {\n  download.file(paste0(base, "5.4/", file), file)\n}\n\n# Set input/output files\nfile_in_hs_uniq <- "fruit_hs_uniq.tsv"\nfile_in_hs_stat <- "fruit_hs_stat.tsv"\nfile_out <- "fruit_hist_uniq_prob.png"\n\n# Data\nhs_uniq <- read.table(file_in_hs_uniq, header = TRUE)\ndata_uniq <- hs_uniq[hs_uniq$k %in% c(5:10), ]\n\nhs_stat <- read.table(file_in_hs_stat, header = TRUE)\ndata_stat <- hs_stat[hs_stat$k %in% c(2:10), ]\n\n# Plot\na <- ggplot(data_uniq, aes(x = unqRat, y = factor(k), fill = factor(k))) +\n  scale_x_continuous(breaks = seq(0.0, 1.0, 0.2)) +\n  xlab("Uniqueness ratio") +\n  ylab("k-mer size") +\n  geom_density_ridges(scale = 6.5, alpha = 0.92) +\n  theme(legend.position = "none")\n\nb <- ggplot(data_stat, aes(x = k, y = unqRat)) +\n  geom_line(color = "darkblue") +\n  geom_point(color = "darkblue", size = 3) +\n  scale_x_continuous(breaks = c(2:10)) +\n  xlab("k-mer size") +\n  ylab("Uniqueness ratio")\n\nc <- ggplot(data_stat, aes(x = k, y = 1 / 21^k)) +\n  geom_line(color = "darkred") +\n  geom_point(color = "darkred", size = 3) +\n  scale_x_continuous(breaks = c(2:10)) +\n  xlab("k-mer size") +\n  ylab(expression(Probability ~ (1 / 21^k)))\n\n# Save plot\nplot <- ggarrange(\n  a,\n  ggarrange(b, c, nrow = 2, align = "v", labels = c("b", "c")),\n  labels = "a",\n  widths = 5:3\n)\nggsave(file_out, plot = plot)\n'})}),"\n",(0,a.jsx)(n.h3,{id:"code-walkthrough-5",children:"Code Walkthrough"}),"\n",(0,a.jsxs)(n.ol,{children:["\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.strong,{children:"Data Loading and Filtering:"})}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:["The datasets ",(0,a.jsx)(n.code,{children:"fruit_hs_uniq.tsv"})," and ",(0,a.jsx)(n.code,{children:"fruit_hs_stat.tsv"})," are downloaded and read."]}),"\n",(0,a.jsxs)(n.li,{children:["Only specific k-mer sizes are retained: k-mer sizes of 5 to 10 for the ",(0,a.jsx)(n.code,{children:"data_uniq"})," dataset and 2 to 10 for ",(0,a.jsx)(n.code,{children:"data_stat"}),"."]}),"\n"]}),"\n"]}),"\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.strong,{children:"Panel A: Density Plot of Uniqueness Ratios by k-mer Size"})}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:["This plot (",(0,a.jsx)(n.code,{children:"a"}),") uses ",(0,a.jsx)(n.code,{children:"geom_density_ridges"})," to show the distribution of the uniqueness ratio (",(0,a.jsx)(n.code,{children:"unqRat"}),") for each k-mer size (from 5 to 10)."]}),"\n",(0,a.jsx)(n.li,{children:"The x-axis represents the uniqueness ratio, and the y-axis shows k-mer sizes."}),"\n",(0,a.jsx)(n.li,{children:"Each k-mer size has its own density ridge, colored by its k value."}),"\n"]}),"\n"]}),"\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.strong,{children:"Panel B: Line Plot of Uniqueness Ratio by k-mer Size"})}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:["This plot (",(0,a.jsx)(n.code,{children:"b"}),") shows how the uniqueness ratio changes with different k-mer sizes."]}),"\n",(0,a.jsxs)(n.li,{children:["A line plot with points shows the progression from ",(0,a.jsxs)(n.span,{className:"katex",children:[(0,a.jsx)(n.span,{className:"katex-mathml",children:(0,a.jsx)(n.math,{xmlns:"http://www.w3.org/1998/Math/MathML",children:(0,a.jsxs)(n.semantics,{children:[(0,a.jsxs)(n.mrow,{children:[(0,a.jsx)(n.mi,{children:"k"}),(0,a.jsx)(n.mo,{children:"="}),(0,a.jsx)(n.mn,{children:"2"})]}),(0,a.jsx)(n.annotation,{encoding:"application/x-tex",children:"k = 2"})]})})}),(0,a.jsxs)(n.span,{className:"katex-html","aria-hidden":"true",children:[(0,a.jsxs)(n.span,{className:"base",children:[(0,a.jsx)(n.span,{className:"strut",style:{height:"0.6944em"}}),(0,a.jsx)(n.span,{className:"mord mathnormal",style:{marginRight:"0.03148em"},children:"k"}),(0,a.jsx)(n.span,{className:"mspace",style:{marginRight:"0.2778em"}}),(0,a.jsx)(n.span,{className:"mrel",children:"="}),(0,a.jsx)(n.span,{className:"mspace",style:{marginRight:"0.2778em"}})]}),(0,a.jsxs)(n.span,{className:"base",children:[(0,a.jsx)(n.span,{className:"strut",style:{height:"0.6444em"}}),(0,a.jsx)(n.span,{className:"mord",children:"2"})]})]})]})," to ",(0,a.jsxs)(n.span,{className:"katex",children:[(0,a.jsx)(n.span,{className:"katex-mathml",children:(0,a.jsx)(n.math,{xmlns:"http://www.w3.org/1998/Math/MathML",children:(0,a.jsxs)(n.semantics,{children:[(0,a.jsxs)(n.mrow,{children:[(0,a.jsx)(n.mi,{children:"k"}),(0,a.jsx)(n.mo,{children:"="}),(0,a.jsx)(n.mn,{children:"10"})]}),(0,a.jsx)(n.annotation,{encoding:"application/x-tex",children:"k = 10"})]})})}),(0,a.jsxs)(n.span,{className:"katex-html","aria-hidden":"true",children:[(0,a.jsxs)(n.span,{className:"base",children:[(0,a.jsx)(n.span,{className:"strut",style:{height:"0.6944em"}}),(0,a.jsx)(n.span,{className:"mord mathnormal",style:{marginRight:"0.03148em"},children:"k"}),(0,a.jsx)(n.span,{className:"mspace",style:{marginRight:"0.2778em"}}),(0,a.jsx)(n.span,{className:"mrel",children:"="}),(0,a.jsx)(n.span,{className:"mspace",style:{marginRight:"0.2778em"}})]}),(0,a.jsxs)(n.span,{className:"base",children:[(0,a.jsx)(n.span,{className:"strut",style:{height:"0.6444em"}}),(0,a.jsx)(n.span,{className:"mord",children:"10"})]})]})]}),"."]}),"\n",(0,a.jsx)(n.li,{children:"The dark blue line represents the uniqueness ratio, a key metric for comparing k-mer uniqueness across sizes."}),"\n"]}),"\n"]}),"\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.strong,{children:"Panel C: Probability Plot by k-mer Size"})}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:["The last plot (",(0,a.jsx)(n.code,{children:"c"}),") uses a red line to show the theoretical probability of a unique k-mer sequence as ",(0,a.jsxs)(n.span,{className:"katex",children:[(0,a.jsx)(n.span,{className:"katex-mathml",children:(0,a.jsx)(n.math,{xmlns:"http://www.w3.org/1998/Math/MathML",children:(0,a.jsxs)(n.semantics,{children:[(0,a.jsxs)(n.mrow,{children:[(0,a.jsx)(n.mn,{children:"1"}),(0,a.jsx)(n.mi,{mathvariant:"normal",children:"/"}),(0,a.jsx)(n.mn,{children:"2"}),(0,a.jsxs)(n.msup,{children:[(0,a.jsx)(n.mn,{children:"1"}),(0,a.jsx)(n.mi,{children:"k"})]})]}),(0,a.jsx)(n.annotation,{encoding:"application/x-tex",children:"1/21^k"})]})})}),(0,a.jsx)(n.span,{className:"katex-html","aria-hidden":"true",children:(0,a.jsxs)(n.span,{className:"base",children:[(0,a.jsx)(n.span,{className:"strut",style:{height:"1.0991em",verticalAlign:"-0.25em"}}),(0,a.jsx)(n.span,{className:"mord",children:"1/2"}),(0,a.jsxs)(n.span,{className:"mord",children:[(0,a.jsx)(n.span,{className:"mord",children:"1"}),(0,a.jsx)(n.span,{className:"msupsub",children:(0,a.jsx)(n.span,{className:"vlist-t",children:(0,a.jsx)(n.span,{className:"vlist-r",children:(0,a.jsx)(n.span,{className:"vlist",style:{height:"0.8491em"},children:(0,a.jsxs)(n.span,{style:{top:"-3.063em",marginRight:"0.05em"},children:[(0,a.jsx)(n.span,{className:"pstrut",style:{height:"2.7em"}}),(0,a.jsx)(n.span,{className:"sizing reset-size6 size3 mtight",children:(0,a.jsx)(n.span,{className:"mord mathnormal mtight",style:{marginRight:"0.03148em"},children:"k"})})]})})})})})]})]})})]}),", given 21 unique values."]}),"\n",(0,a.jsx)(n.li,{children:"The probability axis on the y-axis decreases exponentially as k increases, reflecting the rarity of unique sequences with larger k-mer sizes."}),"\n"]}),"\n"]}),"\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.strong,{children:"Arrangement and Saving"})}),"\n"]}),"\n"]}),"\n",(0,a.jsx)(n.h2,{id:"plot-extra",children:"Plot extra"}),"\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.img,{src:s(62662).Z+"",width:"2100",height:"2400"})}),"\n",(0,a.jsx)(n.pre,{children:(0,a.jsx)(n.code,{className:"language-r",metastring:'title="R"',children:'library(ggplot2)\nlibrary(ggpubr)\nlibrary(dplyr)\n\n# Set theme\ntheme_set(theme_bw())\n\n# Download datasets\nbase <- "https://raw.githubusercontent.com/smortezah/phd/refs/heads/main/plot/"\ndownload.file(\n  paste0(base, "extra/survey_comp_ratio.tsv"),\n  "survey_comp_ratio.tsv"\n)\n\n# Set input/output files\nfile_in <- "survey_comp_ratio.tsv"\nfile_out <- "survey_comp_ratio.png"\n\n# Data\ndf <- read.table(file_in, header = TRUE)\ndf_ref_free <- df %>% filter(format != ".")\ndf_ref_based <- df %>% filter(format == ".")\n\n# Plot\na <- ggplot(df_ref_free, aes(x = order, y = comp.ratio, color = method)) +\n  geom_segment(\n    aes(x = order, xend = order, y = 0.0, yend = comp.ratio),\n    linewidth = 0.75\n  ) +\n  geom_point(size = 10) +\n  geom_text(label = df_ref_free$comp.ratio, color = "white", size = 2.9) +\n  coord_flip() +\n  ggtitle("Reference-free compression") +\n  ylab("Compression ratio") +\n  scale_x_discrete(breaks = df_ref_free$order, labels = df_ref_free$method) +\n  facet_grid(\n    factor(df_ref_free$format, levels = c("FASTA", "FASTQ", "BAM")) ~ .,\n    scales = "free_y", space = "free_y"\n  ) +\n  theme(axis.title.y = element_blank(), legend.position = "none")\n\nb <- ggplot(df_ref_based, aes(x = order, y = comp.ratio, color = method)) +\n  geom_segment(\n    aes(x = order, xend = order, y = 0.0, yend = comp.ratio),\n    linewidth = 0.75\n  ) +\n  geom_point(size = 10) +\n  geom_text(label = df_ref_based$comp.ratio, color = "white", size = 2.9) +\n  coord_flip() +\n  ggtitle("Reference-based compression") +\n  ylab("Compression ratio") +\n  scale_x_discrete(\n    breaks = df_ref_based$order,\n    labels = c("GeCo", "      iDoComp", "GDC 2", "GReEn", "ERGC")\n  ) +\n  theme(axis.title.y = element_blank(), legend.position = "none")\n\n# Save plot\nplot <- ggarrange(a, b, labels = c("a", "b"), nrow = 2, heights = c(3, 1.2))\nggsave(file_out, plot = plot, height = 8)\n'})}),"\n",(0,a.jsx)(n.h3,{id:"code-walkthrough-6",children:"Code walkthrough"}),"\n",(0,a.jsxs)(n.ol,{children:["\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.strong,{children:"Data Loading and Filtering:"})}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:["The dataset ",(0,a.jsx)(n.code,{children:"survey_comp_ratio.tsv"})," is downloaded, read, and filtered into two subsets: ",(0,a.jsx)(n.code,{children:"df_ref_free"}),", representing reference-free compression, and ",(0,a.jsx)(n.code,{children:"df_ref_based"}),", representing reference-based compression."]}),"\n"]}),"\n"]}),"\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.strong,{children:"Panel A: Reference-Free Compression Ratios"})}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:["This plot (",(0,a.jsx)(n.code,{children:"a"}),") shows compression ratios across three file formats (FASTA, FASTQ, BAM) for various reference-free methods."]}),"\n",(0,a.jsx)(n.li,{children:"Each point represents a method's compression ratio, with colored segments extending from the y-axis to the ratio values."}),"\n",(0,a.jsx)(n.li,{children:"Points are labeled with the compression ratio values, formatted as text inside each point."}),"\n",(0,a.jsxs)(n.li,{children:["The x-axis represents the compression ratio, while methods are displayed along the y-axis in a horizontal layout, ordered by the ",(0,a.jsx)(n.code,{children:"order"})," column."]}),"\n"]}),"\n"]}),"\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.strong,{children:"Panel B: Reference-Based Compression Ratios"})}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:["This plot (",(0,a.jsx)(n.code,{children:"b"}),") displays compression ratios for reference-based methods, with a similar layout to panel A."]}),"\n",(0,a.jsx)(n.li,{children:'The x-axis shows compression ratios, while the y-axis lists reference-based methods (including methods like "GeCo," "iDoComp," "GDC 2," etc.).'}),"\n",(0,a.jsx)(n.li,{children:"Compression ratio values are labeled within each point, using white text for visibility."}),"\n"]}),"\n"]}),"\n",(0,a.jsxs)(n.li,{children:["\n",(0,a.jsx)(n.p,{children:(0,a.jsx)(n.strong,{children:"Arrangement and Saving:"})}),"\n",(0,a.jsxs)(n.ul,{children:["\n",(0,a.jsxs)(n.li,{children:["The plots are arranged vertically, with reference-free compression (panel ",(0,a.jsx)(n.code,{children:"a"}),") at the top and reference-based compression (panel ",(0,a.jsx)(n.code,{children:"b"}),") below."]}),"\n",(0,a.jsxs)(n.li,{children:["The layout is saved as a PNG file (",(0,a.jsx)(n.code,{children:"survey_comp_ratio.png"}),"), providing a detailed comparison between the two compression approaches."]}),"\n"]}),"\n"]}),"\n"]})]})}function h(e={}){const{wrapper:n}={...(0,t.a)(),...e.components};return n?(0,a.jsx)(n,{...e,children:(0,a.jsx)(d,{...e})}):d(e)}},52947:(e,n,s)=>{s.d(n,{Z:()=>a});const a=s.p+"assets/images/ggplot2-ac_nc-2fc9077e0d719b30ae926e01f1d7f0ed.png"},81455:(e,n,s)=>{s.d(n,{Z:()=>a});const a=s.p+"assets/images/ggplot2-cryfa_enc_aes-3863a63810eba1ffe90a265e7754c4b0.png"},87197:(e,n,s)=>{s.d(n,{Z:()=>a});const a=s.p+"assets/images/ggplot2-cryfa_redun-44b24f05913c3c0cb97ced83666732ea.png"},64801:(e,n,s)=>{s.d(n,{Z:()=>a});const a=s.p+"assets/images/ggplot2-cryfa_thread-e79407bba8283e77a6fc85385b5f3bef.png"},88847:(e,n,s)=>{s.d(n,{Z:()=>a});const a=s.p+"assets/images/ggplot2-fruit_hist_uniq_prob-72efb82880a72a807b7a2dbca0dcdd21.png"},56761:(e,n,s)=>{s.d(n,{Z:()=>a});const a=s.p+"assets/images/ggplot2-ir_hs_gg-d22f401307f356f66a7c74d1ae51711f.png"},62662:(e,n,s)=>{s.d(n,{Z:()=>a});const a=s.p+"assets/images/ggplot2-survey_comp_ratio-a76d4779363aa9c7f76947af4d610be5.png"},11151:(e,n,s)=>{s.d(n,{Z:()=>l,a:()=>r});var a=s(67294);const t={},i=a.createContext(t);function r(e){const n=a.useContext(i);return a.useMemo((function(){return"function"==typeof e?e(n):{...n,...e}}),[n,e])}function l(e){let n;return n=e.disableParentContext?"function"==typeof e.components?e.components(t):e.components||t:r(e.components),a.createElement(i.Provider,{value:n},e.children)}}}]);